### PR TITLE
Use thread-safe cache in thread_local_caching_allocator

### PR DESCRIPTION
**Pull request by Pansysk75**

### CODE REVIEW

**Review by hkaiser — DISMISSED** [review]
The remaining problems are unrelated. LGTM! Thanks!

File: libs/core/allocator_support/include/hpx/allocator_support/thread_local_caching_allocator.hpp [hunk 1]
```diff
@@ -82,8 +83,9 @@ namespace hpx::util {
                 }
                 else
                 {
-                    p = data.top().first;
-                    data.pop();
+                    std::pair<T*, size_type> pair;
+                    data.pop(pair);
```
Comment (hkaiser) [inline libs/core/allocator_support/include/hpx/allocator_support/thread_local_caching_allocator.hpp:87]: I think you should check for the return value here as between the check `data.empty()` above and here somebody could have removed the item. Even better if you rewrote the whole functions to something like:
```
std::pair<T*, size_type> pair;
if (!data.pop(pair)) {
    // allocate new
}
else {
   p = pair.first;
}
```

Comment (Pansysk75) [reply libs/core/allocator_support/include/hpx/allocator_support/thread_local_caching_allocator.hpp]: Fixed, thanks!

**Comment by hkaiser** [comment]
@Pansysk75 we still have the circular dependency between the concurrency and allocator modules. Also, some of the header tests are failing (most likely caused by the circular dependency issue)

**Comment by Pansysk75** [comment]
> @Pansysk75 we still have the circular dependency between the concurrency and allocator modules. Also, some of the header tests are failing (most likely caused by the circular dependency issue)

Hmm I suppose that means the `allocator_support` module is higher in the module hierarchy than the `concurrency` module?
I'm not exactly sure what fixing this would involve.

**Comment by hkaiser** [comment]
> > @Pansysk75 we still have the circular dependency between the concurrency and allocator modules. Also, some of the header tests are failing (most likely caused by the circular dependency issue)
> 
> Hmm I suppose that means the `allocator_support` module is higher in the module hierarchy than the `concurrency` module? I'm not exactly sure what fixing this would involve.

Yes, see: https://hpx-docs.stellar-group.org/latest/report/v1.10.0/core/allocator_support.html

> Oh, also, we should probably make all accesses to members of `allocated_cache` thread safe... Not sure why I ignored that. Also, this now makes me dislike this solution even more.

Yes, the `deallocated` and `allocated` members should be atomic. But we should combine both variables into one counter for that.

> I opened an issue #6540 to document this.

Thanks!

**Comment by hkaiser** [comment]
@Pansysk75 I think this is good to go as a stop-gap measure circumventing the underlying problem from happening. What do you think?

**Comment by Pansysk75** [comment]
> @Pansysk75 I think this is good to go as a stop-gap measure circumventing the underlying problem from happening. What do you think?

Sure, one last thing, what is this check supposed to do? Our underlying data structure has a .size() member function, is it any different to use that?

**Comment by hkaiser** [comment]
@Pansysk75 the original issues are resolved here. Let's merge it now.

**Pull request merged**